pkgs/development/pharo/vm/build-vm.nix

```diff
@@ -1,76 +1,103 @@
-    set -f
+  # Configure with options modeled on the 'mvm' build script from the vm.
+  configureScript = "platforms/unix/config/configure";
+  configureFlags = [ "--without-npsqueak"
+                     "--with-vmversion=5.0"
+                     "--with-src=${vm}" ];
+  CFLAGS = "-msse2 -D_GNU_SOURCE -DCOGMTVM=0 -g -O2 -DNDEBUG -DDEBUGVM=0";
+  LDFLAGS = "-Wl,-z,now";
+
+  # VM sources require some patching before build.
+  prePatch = ''
+    patchShebangs opensmalltalk-vm/build.${flavor}
+    # Fix hard-coded path to /bin/rm in a script
+    sed -i -e 's:/bin/rm:rm:' opensmalltalk-vm/platforms/unix/config/mkmf
+    # Fill in mandatory metadata about the VM source version
+    sed -i -e 's!\$Date\$!$Date: ${source-date} $!' \
+           -e 's!\$Rev\$!$Rev: ${version} $!' \
+           -e 's!\$URL\$!$URL: ${source-url} $!' \
+           opensmalltalk-vm/platforms/Cross/vm/sqSCCSVersion.h
+  '';
 
-    exec $prefix/lib/$name/pharo-vm "\$@"
-    EOF
+  # Note: --with-vmcfg configure option is broken so copy plugin specs to ./
+  preConfigure = ''
+    cd opensmalltalk-vm
+    cp build.${flavor}/pharo.cog.spur/plugins.{ext,int} .
+  '';
 
-    cat > $prefix/bin/${binary-basename}-nox <<EOF
-    #!${bash}/bin/bash
+  # (No special build phase.)
 
-    # disable parameter expansion to forward all arguments unprocessed to the VM
+  installPhase = ''
+    # Install in working directory and then copy
+    make install-squeak install-plugins prefix=$(pwd)/products
+
+    # Copy binaries & rename from 'squeak' to 'pharo'
+    mkdir -p $out
+    cp    products/lib/squeak/5.0-*/squeak $out/pharo
+
+    cp -r products/lib/squeak/5.0-*/*.so $out
+    ln -s "${pharo-share}/lib/"*.sources $out
+
+    # Create a shell script to run the VM in the proper environment.
+    # 
+    # These wrapper puts all relevant libraries into the
+    # LD_LIBRARY_PATH. This is important because various C code in the VM
+    # and Smalltalk code in the image will search for them there.
+    mkdir -p $out/bin
+    chmod u+w $out/bin
+
+    # Note: include ELF rpath in LD_LIBRARY_PATH for finding libc.
+    libs=$out:$(patchelf --print-rpath $out/pharo):${cairo}/lib:${mesa}/lib:${freetype}/lib:${openssl}/lib:${libuuid}/lib:${alsaLib}/lib:${xorg.libICE}/lib:${xorg.libSM}/lib
+
+    # Create the script
+    cat > $out/bin/${cmd} <<EOF
+    #!/bin/sh
     set -f
-
-    exec $prefix/lib/$name/pharo-vm -vm-display-null "\$@"
+    LD_LIBRARY_PATH="\$LD_LIBRARY_PATH:$libs" exec $out/pharo "\$@"
     EOF
+    chmod +x $prefix/bin/${cmd}
+  '';
 
-    chmod +x $prefix/bin/${binary-basename}-x $prefix/bin/${binary-basename}-nox
+  enableParallelBuilding = true;
 
-    ln -s "${pharo-share}/lib/"*.sources $prefix/lib/$name
-  '';
+  # Note: Force gcc6 because gcc5 crashes when compiling the VM.
+  buildInputs = [ bash unzip glibc openssl gcc6 mesa freetype xorg.libX11 xorg.libICE xorg.libSM alsaLib cairo pharo-share libuuid autoreconfHook ];
```

Comment:
Some of these are only build-time dependencies, e.g. unzip, gcc. Those should be listed as `nativeBuildInputs` instead, since `buildInputs` is meant for runtime dependencies